tracing-opentelemetry/src/layer.rs

```diff
@@ -192,9 +197,72 @@ impl<'a> field::Visit for SpanEventVisitor<'a> {
             }
         }
     }
+
+    /// Set attributes on the underlying OpenTelemetry [`Span`] using a [`std::error::Error`]'s
+    /// [`std::fmt::Display`] implementation. Also adds the `source` chain as an extra field
+    ///
+    /// [`Span`]: opentelemetry::trace::Span
+    fn record_error(
+        &mut self,
+        field: &tracing_core::Field,
+        value: &(dyn std::error::Error + 'static),
+    ) {
+        let mut chain = Vec::new();
+        let mut next_err = value.source();
+
+        while let Some(err) = next_err {
+            chain.push(Cow::Owned(err.to_string()));
+            next_err = err.source();
+        }
+
+        if self.2.record {
+            self.0
+                .attributes
+                .push(Key::new("exception.message").string(value.to_string()));
+
+            // NOTE: This is actually not the stacktrace of the exception. This is
+            // the "source chain". It represents the heirarchy of errors from the
+            // app level to the lowest level such as IO. It does not represent all
+            // of the callsites in the code that led to the error happening.
+            // `std::error::Error::backtrace` is a nightly-only API and cannot be
+            // used here until the feature is stabilized.
+            self.0
+                .attributes
+                .push(Key::new("exception.stacktrace").array(chain.clone()));
+        }
+
+        if self.2.propagate {
+            if let Some(span) = &mut self.1 {
+                if let Some(attrs) = span.attributes.as_mut() {
+                    attrs.push(Key::new("exception.message").string(value.to_string()));
```

Comment:
since the field names are repeated a few times now, i might consider making them constants?